### PR TITLE
Full rewrite

### PR DESCRIPTION
**Pull request by rv32ima**
This is a PR tracking my status on rewriting Tarmac. Most of the code here will get touched - frankly a lot of it is inadequate, and not up to modern standards. I'm planning on keeping the user-facing API the same, just simplifying a lot of the code.

### CODE REVIEW

**Comment by RBXTyler** [comment]
CI / build failed? 👀

File: src/roblox_api/open_cloud.rs [hunk 1]
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
...
-                })
-            }
-
-            result => result,
-        }
-    }
-
-    fn upload_image(&mut self, data: &ImageUploadData) -> Result<UploadResponse, RobloxApiError> {
-        self.upload_image_inner(data)
+    // this was a bad idea, sorry
```
Comment (OutOfBears) [inline src/roblox_api/open_cloud.rs]: Should we remove this?

File: src/roblox_api/open_cloud.rs [hunk 2]
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
...
+
+    async fn upload_image(&self, data: ImageUploadData<'a>) -> Result<UploadResponse> {
+        self.upload_image_inner(data).await
     }
 
-    fn download_image(&mut self, id: u64) -> Result<Vec<u8>, RobloxApiError> {
-        LegacyClient::new(self.credentials.clone())?.download_image(id)
+    fn download_image(&self, id: u64) -> Result<Vec<u8>> {
+        todo!();
+        // LegacyClient::new(self.credentials.clone())?.download_image(id)
```
Comment (OutOfBears) [inline src/roblox_api/open_cloud.rs]: Should we remove this function?

File: src/sync_backend.rs [hunk 3]
```diff
@@ -32,48 +38,65 @@ pub struct UploadInfo {
...
 
-impl<'a> RobloxSyncBackend<'a> {
-    pub fn new(api_client: &'a mut dyn RobloxApiClient) -> Self {
-        Self { api_client }
+impl<'a, ApiClient> RobloxSyncBackend<'a, ApiClient>
+where
+    ApiClient: RobloxApiClient<'a> + Sync + Clone + Send,
+{
+    pub fn new(api_client: ApiClient) -> Self {
+        // Self { api_client: Arc::new(api_client) }
```
Comment (OutOfBears) [inline src/sync_backend.rs]: \>:(

File: src/sync_backend.rs [hunk 4]
```diff
@@ -32,48 +38,65 @@ pub struct UploadInfo {
...
-            Err(RobloxApiError::ResponseError {
-                status: StatusCode::TOO_MANY_REQUESTS,
-                ..
-            }) => Err(Error::RateLimited),
-
-            Err(err) => Err(err.into()),
-        }
+        todo!()
+
+        // let result = self
```
Comment (OutOfBears) [inline src/sync_backend.rs]: \>:(

File: src/sync_backend.rs [hunk 5]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
...
 
-        let file_path = path.join(id.to_string());
-        fs::write(&file_path, &data.contents)?;
+        // let file_path = path.join(id.to_string());
+        // fs::write(&file_path, &data.contents)?;
 
-        Ok(UploadResponse {
-            id: AssetId::Id(id),
-        })
+        // Ok(UploadResponse {
```
Comment (OutOfBears) [inline src/sync_backend.rs]: \>:(

File: src/sync_backend.rs [hunk 6]
```diff
@@ -234,116 +264,116 @@ pub enum Error {
...
-
-            assert_eq!(counter, 3);
-            assert!(match upload_result {
-                Error::RateLimited => true,
-                _ => false,
-            });
-        }
-    }
-}
+// #[cfg(test)]
```
Comment (OutOfBears) [inline src/sync_backend.rs]: \>:(

File: src/roblox_api/open_cloud.rs (hunk 1 above)
Comment (RBXTyler) [reply src/roblox_api/open_cloud.rs]: We need that method? Because the image will not upload if it is moderated.

Comment (OutOfBears) [reply src/roblox_api/open_cloud.rs]: The only reason it ever gets moderated is for name, so we could hash it?

Comment (RBXTyler) [reply src/roblox_api/open_cloud.rs:53]: >:(

Comment (RBXTyler) [reply src/roblox_api/open_cloud.rs]: \>:(

File: src/main.rs [hunk 7]
```diff
@@ -14,26 +14,31 @@ mod sync_backend;
...
     match options.command {
-        Subcommand::UploadImage(upload_options) => {
-            commands::upload_image(options.global, upload_options)?
+        Command::UploadImage(upload_options) => {
+            commands::upload_image(options.global, upload_options).await?
         }
-        Subcommand::Sync(sync_options) => commands::sync(options.global, sync_options)?,
-        Subcommand::CreateCacheMap(sub_options) => {
+        Command::Sync(sync_options) => {
+            // commands::sync(options.global, sync_options)?,
```
Comment (OutOfBears) [inline src/main.rs]:
```suggestion
```

File: src/roblox_api/mod.rs [hunk 8]
```diff
@@ -33,19 +34,19 @@ pub struct RobloxCredentials {
...
-        Self: Sized;
+#[async_trait]
+pub trait RobloxApiClient<'a> {
+    fn new(credentials: RobloxCredentials) -> Result<Self>
+    where Self: Sized;
+    // this was a bad idea, sorry
+    // async fn upload_image_with_moderation_retry(
+    //     &self,
+    //     data: ImageUploadData<'a>,
+    // ) -> Result<UploadResponse>;
```
Comment (OutOfBears) [inline src/roblox_api/mod.rs]:
```suggestion
```

File: src/roblox_api/mod.rs [hunk 9]
```diff
@@ -129,7 +130,8 @@ Tarmac will attempt to upload to the user currently logged into Roblox Studio, o
 If you mean to use the Open Cloud API, make sure to provide an API key!")
             };
 
-            Ok(Box::new(LegacyClient::new(credentials)?))
+            todo!();
+            // Ok(Box::new(LegacyClient::new(credentials)?))
```
Comment (OutOfBears) [inline src/roblox_api/mod.rs]:
```suggestion
```

File: src/roblox_api/open_cloud.rs [hunk 10]
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
...
+    //             self.upload_image(ImageUploadData {
+    //                 name: "image".to_string(),
+    //                 ..data.to_owned()
+    //             })
+    //             .await
+    //         }
+
+    //         result => result,
+    //     }
+    // }
```
Comment (OutOfBears) [inline src/roblox_api/open_cloud.rs]:
```suggestion
```

File: src/roblox_api/open_cloud.rs (hunk 2 above)
Comment (OutOfBears) [inline src/roblox_api/open_cloud.rs]:
```suggestion
```

File: src/sync_backend.rs (hunk 3 above)
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion
```

File: src/sync_backend.rs [hunk 11]
```diff
@@ -32,48 +38,65 @@ pub struct UploadInfo {
...
+        //         })
+        //     }
+
+        //     Err(RobloxApiError::ResponseError {
+        //         status: StatusCode::TOO_MANY_REQUESTS,
+        //         ..
+        //     }) => Err(Error::RateLimited),
+
+        //     Err(err) => Err(err.into()),
+        // }
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion
```

File: src/sync_backend.rs [hunk 12]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
     }
 }
 
+#[async_trait]
 impl SyncBackend for DebugSyncBackend {
-    fn upload(&mut self, data: UploadInfo) -> Result<UploadResponse, Error> {
-        log::info!("Copying {} to local folder", &data.name);
+    async fn upload(&self, data: UploadInfo) -> Result<UploadResponse, Error> {
+        todo!();
+        // log::info!("Copying {} to local folder", &data.name);
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion
```

File: src/sync_backend.rs [hunk 13]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
...
 
-        self.last_id += 1;
-        let id = self.last_id;
+        // self.last_id += 1;
+        // let id = self.last_id;
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion

```

File: src/sync_backend.rs [hunk 14]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
...
 
-        let path = Path::new(".tarmac-debug");
-        fs::create_dir_all(path)?;
+        // let path = Path::new(".tarmac-debug");
+        // fs::create_dir_all(path)?;
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion

```

File: src/sync_backend.rs [hunk 15]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
...
 
-        let file_path = path.join(id.to_string());
-        fs::write(&file_path, &data.contents)?;
+        // let file_path = path.join(id.to_string());
+        // fs::write(&file_path, &data.contents)?;
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion

```

File: src/sync_backend.rs [hunk 16]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
...
 
-        Ok(UploadResponse {
-            id: AssetId::Id(id),
-        })
+        // Ok(UploadResponse {
+        //     id: AssetId::Id(id),
+        // })
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion

```

File: src/sync_backend.rs [hunk 17]
```diff
@@ -234,116 +264,116 @@ pub enum Error {
...
+//             let upload_result = backend.upload(any_upload_info()).unwrap_err();
+
+//             assert_eq!(counter, 3);
+//             assert!(match upload_result {
+//                 Error::RateLimited => true,
+//                 _ => false,
+//             });
+//         }
+//     }
+// }
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion

```

File: src/sync_backend.rs [hunk 18]
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
...
+        // Ok(UploadResponse {
+        //     id: AssetId::Id(id),
+        // })
     }
 }
 
 /// Performs the retry logic for rate limitation errors. The struct wraps a SyncBackend so that
 /// when a RateLimited error occurs, the thread sleeps for a moment and then tries to reupload the
 /// data.
+///
```
Comment (OutOfBears) [inline src/sync_backend.rs]:
```suggestion
```

**Review by OutOfBears — CHANGES_REQUESTED** [review]

File: src/roblox_api/open_cloud.rs [hunk 19]
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
...
+
+    async fn upload_image(&self, data: ImageUploadData<'a>) -> Result<UploadResponse> {
+        self.upload_image_inner(data).await
     }
 
-    fn download_image(&mut self, id: u64) -> Result<Vec<u8>, RobloxApiError> {
-        LegacyClient::new(self.credentials.clone())?.download_image(id)
+    fn download_image(&self, id: u64) -> Result<Vec<u8>> {
+        todo!();
```
Comment (RBXTyler) [inline src/roblox_api/open_cloud.rs]: ????????????

**Pull request closed without merging**